src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -799,6 +807,24 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded, Text
                                         this.ProjectState.AnalyzerOptions.AnalyzerConfigOptionsProvider,
                                         additionalTexts));
                             }
+                            else
+                            {
+#if DEBUG
+
+                                // Assert that the generator driver is in sync with our additional document states; there's not a public
+                                // API to get this, but we'll reflect in DEBUG-only.
+                                var driverType = generatorInfo.Driver.GetType();
+                                var stateMember = driverType.GetField("_state", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+                                Contract.ThrowIfNull(stateMember);
+                                var additionalTextsMember = stateMember.FieldType.GetField("AdditionalTexts", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+                                Contract.ThrowIfNull(additionalTextsMember);
+                                var state = stateMember.GetValue(generatorInfo.Driver);
+                                var additionalTexts = (ImmutableArray<AdditionalText>)additionalTextsMember.GetValue(state)!;
+
+                                Contract.ThrowIfFalse(additionalTexts.Length == this.ProjectState.AdditionalDocumentStates.Count);
```

Comment:
FYI to @chsienki. :smile:

Reply:
(and yes, it's DEBUG only)

Reply:
![image](https://user-images.githubusercontent.com/16246502/142055246-88902307-b8b9-4a9b-8b60-48bd2e472045.png)

More seriously though, should we just expose this / state in general?

Reply:
@chsienki So I haven't had a need to see this state for a *functional* reason, but it's sure helpful when debugging issues so I can add better asserts like this.

Reply:
Chatting with @chsienki, I entirely agree with his meme, but since it's DEBUG-only code we're going to leave it in, and we'll discuss adding better APIs separately.